[Console] fix: console deinit always deadlocks - adds ability to unblock linenoise (try 2) (IDFGH-9188)

chipweinberger opened the pull request:
**Previous PR:** https://github.com/espressif/esp-idf/pull/9983

I am re-opening this PR to get more traction on it. 

This solves: https://github.com/espressif/esp-idf/issues/9974

**Motivation:**
- currently, console deinit causes a deadlock
- without deinit, we can't create consoles "as needed"
- for example, USB Serial JTAG console should be destroyed when entering USB host mode
- for example, "web" consoles should be destroyed when the browser window is closed

chipweinberger commented on components/esp_ringbuf/include/freertos/ringbuf.h:
```diff
@@ -496,6 +496,16 @@ void vRingbufferGetInfo(RingbufHandle_t xRingbuffer,
+ *
+ * @param[in]   xRingbuffer     The ring buffer who's rx will be unblocked.
+ */
+void vRingbufferUnblockRx(RingbufHandle_t xRingbuffer);
```
> If usb_serial_jtag.c starts using something other than a ringbuffer internally (e.g. a StreamBuffer), we would have to add similar API for unblocking there; and given that StreamBuffer is upstream FreeRTOS code we would most likely want to avoid that.

^ From Ivan. 

Can you re-evaluate your position on this? `vRingbufferUnblockRx` seems like the ideal solution.

If we ever switched to `StreamBuffer` we could come up with alternative solutions. For example:
- A) a counting semaphore + StreamBuffer. 'give' the semaphore on new data + then call StreamBuffer with a 0 timeout.
- B) Use a long timeout with StreamBuffer, perhaps 10 seconds, and poll an exit condition
- C) Make a Pull Request on FreeRTOS to add the necessary unblock feature to StreamBuffer

chipweinberger replied:
Alternatively, ^^ solution (A), could be used to mitigate the need for `vRingbufferUnblockRx` in this PR as well. 

I think the solution would look like this:
- **normal exclusive mutex.**  `usb_serial_jtag_read_bytes` will grab this mutex first. _Not entirely necessary._ This makes sure the counting semaphore only has a single thread using it at a time. **i.e.** so the first thread to call `read()` will always be satisfied first.
- **counting semaphore.** `usb_serial_jtag_isr_handler_default` increments this counting semaphore _once_ each time it appends to the ringbuffer, `usb_serial_jtag_read_bytes` will 'take' from the semaphore, in order to wait until new bytes are ready (We could also use `event_groups` instead).
- **ring buffer**, `usb_serial_jtag_read_bytes` calls the ringbuffer with a zero timeout, and will wait on the counting semaphore again until there is enough data.

Of course, I would consider the `vRingbufferUnblockRx` solution better.

SoucheSouche commented:
Hi @chipweinberger, I started looking into your PR. Can I ask you to rebase and fix the build issues and conflict with upstream?

chipweinberger commented:
sure, one moment.

chipweinberger commented:
@SoucheSouche updated. was just a trivial collison

SoucheSouche commented on components/esp_ringbuf/ringbuf.c:
```diff
@@ -1351,6 +1361,18 @@ void vRingbufferGetInfo(RingbufHandle_t xRingbuffer,
+    configASSERT(pxRingbuffer);
+
+    // is the semaphore taken?
+    if (uxSemaphoreGetCount(rbGET_RX_SEM_HANDLE(pxRingbuffer)) == 0) {
```
you still need to `#include "freertos/semphr.h"` to use the function.

chipweinberger replied:
updated.

SoucheSouche commented on components/driver/uart/uart.c:
```diff
@@ -1762,6 +1762,16 @@ esp_err_t uart_set_loop_back(uart_port_t uart_num, bool loop_back_en)
+        return ESP_ERR_INVALID_STATE;
+    }
+
+    vRingbufferUnblockRx(pp_uart_obj[uart_num]->rx_ring_buf);
```
you left a typo here: `pp_uart_obj` → `p_uart_obj`

chipweinberger replied:
updated.

SoucheSouche commented on components/esp_ringbuf/ringbuf.c:
```diff
@@ -1351,6 +1362,18 @@ void vRingbufferGetInfo(RingbufHandle_t xRingbuffer,
+    configASSERT(pxRingbuffer);
+
+    // is the semaphore taken?
+    if (uxSemaphoreGetCount(rbGET_RX_SEM_HANDLE(pxRingbuffer)) == 0) {
```
`rbGET_RX_SEM_HANDLE` doesn't seem to be declared anywhere.

chipweinberger replied:
This code must have changed. @SoucheSouche, I'm pretty behind and stressed about my own work at the moment. If you are able to update it that would be a great help.

SoucheSouche replied:
sure no problem!

chipweinberger replied:
I wonder if we should move this from the flags, and into its own variable. That way we are less likely to hit threading problems from multiple flags sharing the same memory location.

chipweinberger replied:
^ I'm not familiar enough with ringbuf to make this judgement call.

SoucheSouche commented on components/console/esp_console_repl.c:
```diff
@@ -409,13 +411,25 @@ static esp_err_t esp_console_repl_uart_delete(esp_console_repl_t *repl)
+    if (repl_com->state == CONSOLE_REPL_STATE_START) {
+        repl_com->state = CONSOLE_REPL_STATE_REPL_STOP_REQUESTED;
+        while (repl_com->state != CONSOLE_REPL_STATE_REPL_TASK_ENDED) {
+            uart_unblock_reads();
```
This function is expecting `uart_num` as a parameter it seems. you should update it to `uart_unblock_reads(uart_repl->uart_channel);`

chipweinberger replied:
updated. I wonder how I was missing that.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/espressif/esp-idf?pullRequest=10580) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/espressif/esp-idf?pullRequest=10580) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/esp-idf?pullRequest=10580) it.</sub>

SoucheSouche commented:
@chipweinberger, there has been some changes in the ring buffer so I am now looking at your PR again since the semaphores for rx and tx were removed.

chipweinberger commented:
> since the semaphores for rx and tx were removed.

I updated the PR and removed the semaphores. But as I was saying I think we should move this bit: `#define rbBUFFER_UNBLOCK_RX_FLAG    ( ( UBaseType_t ) 32 )` to its own variable, to simplify race condition worries.

SoucheSouche commented:
The other issue is that [this comment](https://github.com/espressif/esp-idf/pull/9983#discussion_r1003527773) is still valid and your current solution won't be accepted.

chipweinberger commented:
@SoucheSouche `xTaskAbortDelay` is going to cause a lot of issues IMO, because it does not give a chance for clenup. Memory leaks at minimum, bugs at worse.

chipweinberger commented:
@SoucheSouche , did you understand this comment? "Thinking about this again, select on a pair of file descriptors (stdin and eventfd) is probably the cleanest solution." I'm not sure what that solution would look like.

If you use `xTaskAbortDelay`, make sure the thread is not blocked in user code.

SoucheSouche commented:
Hi @chipweinberger,
Sorry that I left this PR not updated for so long. I will get back at it this week. It's been a while so it might take a bit of time to get back into it but I will update you as I go through it.
Hopefully we will finally be able to merge it this time.

SoucheSouche commented:
Hi @chipweinberger, a little update on the PR.
Since it was fairly old, I first had to rebase it on the latest master and realized that your solution was no longer working.
The concept of adding 2 states to the console repl is great and works just fine but the solution that you were using to force linenoise to return didn't seem to work.

I experimented with `select` as @igrr suggested and ended up finding a solution.
The very drafty logic implemented in `linenoiseDumb()` and `linenoiseRaw()` is the following:
```
[...]
        if (select(in_fd + 1, &read_fds, NULL, &except_fds, NULL) < 0) {
            // select returned with -1, handle the error
        }

        // std_in is either ready to read or an exceptional condition occurred
        if(FD_ISSET(in_fd, &except_fds)) {
            // exceptional condition occurred, return from linenoise back into the while loop
            // of `esp_console_repl_task` where the repl state will be reevaluated
            break;
        }
        else if(FD_ISSET(in_fd, &read_fds)) {
            // do the reading since std_in is ready for reading
        }
        else {
            continue;
        }
```
I kept `uart_unblock_reads` and `usb_serial_jtag_unblock_reads` and I make them fire an "exceptional condition" event that will make the `select` return and `if(FD_ISSET(in_fd, &except_fds))` evaluated to true.

chipweinberger commented:
seems like a reasonable solution!

impressive :) fun :)

SoucheSouche commented:
I had to implement the select for the usb serial jtag driver so I will have to first merge this added feature before proceeding with the actual deadlock fix.
I will keep you posted.

chipweinberger commented:
related: https://github.com/espressif/esp-idf/issues/9964

SoucheSouche commented:
Hi @chipweinberger,
a little update on the fix.
I am still waiting to merge the select() feature for the usb cdc and usb serial jtag drivers before I can merge the fix for the deadlock.
I will let you know when the select() features have been merged but it might take a little while.

SoucheSouche commented:
hi @chipweinberger,
Sorry I didn't update you for so long. It was quite a journey to implement and merge the select functionality in the usb serial jtag driver and the usb cdc driver.
But great news, I just merged the last one today so I can finally update the bugfix and submit it for review.

I will try to submit it before the end of the week, so I hope it will be merged before Christmas.

chipweinberger commented:
Yes I can imagine it was a difficult fix. Congrats!

My product is now shipping: [jamcorder.com](jamcorder.com)

SoucheSouche commented:
Congratulation to you!
The project looks really good! I wish you the best.

And I couldn't help but notice that you are missing the mouse hover on the FAQ section for the desktop version of the website.

chipweinberger commented:
very nice!

SoucheSouche commented:
@chipweinberger it took its time but it is done :)
